File: solutions/erlang/protein-translation/1/src/protein_translation.erl
-module(protein_translation).

-export([proteins/1]).

proteins(Strand) -> proteins(Strand, []).

proteins([], Acc) -> {ok, lists:reverse(Acc)};
proteins([A, B, C | Rest], Acc) ->
    P = protein([A, B, C]),
    case P of
        {error, badarg} -> {error, badarg};
        stop -> {ok, lists:reverse(Acc)};
        _ -> proteins(Rest, [P|Acc])
    end;
proteins(_, _) -> {error, badarg}.

protein("AUG") -> methionine;
protein("UUU") -> phenylalanine;
protein("UUC") -> phenylalanine;
protein("UUA") -> leucine;
protein("UUG") -> leucine;
protein("UCU") -> serine;
protein("UCC") -> serine;
protein("UCA") -> serine;
protein("UCG") -> serine;
protein("UAU") -> tyrosine;
protein("UAC") -> tyrosine;
protein("UGU") -> cysteine;
protein("UGC") -> cysteine;
protein("UGG") -> tryptophan;
protein("UAA") -> stop;
protein("UAG") -> stop;
protein("UGA") -> stop;
protein(_) -> {error, badarg}.

File: solutions/elixir/protein-translation/1/lib/protein_translation.ex
defmodule ProteinTranslation do
  @codon_map %{
    "UGU" => "Cysteine",
    "UGC" => "Cysteine",
    "UUA" => "Leucine",
    "UUG" => "Leucine",
    "AUG" => "Methionine",
    "UUU" => "Phenylalanine",
    "UUC" => "Phenylalanine",
    "UCU" => "Serine",
    "UCC" => "Serine",
    "UCA" => "Serine",
    "UCG" => "Serine",
    "UGG" => "Tryptophan",
    "UAU" => "Tyrosine",
    "UAC" => "Tyrosine",
    "UAA" => "STOP",
    "UAG" => "STOP",
    "UGA" => "STOP"
  }
  @doc """
  Given an RNA string, return a list of proteins specified by codons, in order.
  """
  @spec of_rna(String.t()) :: {:ok, list(String.t())} | {:error, String.t()}
  def of_rna(rna) do
    codons = rna
    |> String.codepoints
    |> Enum.chunk_every(3)
    |> Enum.map(&Enum.join/1)

    protein = Enum.reduce_while(codons, [], fn elem, acc ->
      case of_codon(elem) do
        {:ok, "STOP"} -> {:halt, acc}
        {:ok, new_elem} -> {:cont, [new_elem | acc]}
        {:error, _} = err -> {:halt, err}
      end
    end)

    if is_list(protein) do
      {:ok, Enum.reverse(protein)}
    else
      {:error, "invalid RNA"}
    end

  end

  @doc """
  Given a codon, return the corresponding protein

  UGU -> Cysteine
  UGC -> Cysteine
  UUA -> Leucine
  UUG -> Leucine
  AUG -> Methionine
  UUU -> Phenylalanine
  UUC -> Phenylalanine
  UCU -> Serine
  UCC -> Serine
  UCA -> Serine
  UCG -> Serine
  UGG -> Tryptophan
  UAU -> Tyrosine
  UAC -> Tyrosine
  UAA -> STOP
  UAG -> STOP
  UGA -> STOP
  """
  @spec of_codon(String.t()) :: {:ok, String.t()} | {:error, String.t()}
  def of_codon(codon) do
    polypeptide = @codon_map[codon]
    case polypeptide do
      nil -> {:error, "invalid codon"}
      _ -> {:ok, polypeptide}
    end
  end
end
